wgengine/router: unfork winipcfg-go package, use upstream

alexbrainman opened the pull request:
Use golang.zx2c4.com/wireguard/windows/tunnel/winipcfg
instead of github.com/tailscale/winipcfg-go package.

Updates #760

Signed-off-by: Alex Brainman <alex.brainman@gmail.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/tailscale/tailscale/792)
<!-- Reviewable:end -->

bradfitz commented:
I sent https://github.com/WireGuard/wireguard-windows/pull/9 upstream. Let's get that in first so we're not losing anything with this change. But otherwise it seems fine. Thanks!

Also, please rebase for the go.sum conflict.

bradfitz commented:
Oh, please re-title the commit message & PR here.

Instead of "update github.com/tailscale/winipcfg-go package", say something like: "unfork winipcfg-go package, use upstream"

alexbrainman commented:
> I sent [WireGuard/wireguard-windows#9](https://github.com/WireGuard/wireguard-windows/pull/9) upstream. Let's get that in first so we're not losing anything with this change. But otherwise it seems fine.

I will let you merge this when you are ready.

> Also, please rebase for the go.sum conflict.

I did rebase on main. And then rebuilt exe. Hopefully that will do.

> Oh, please re-title the commit message & PR here.

Done. Both.

Alex

alexbrainman commented:
I also don't know much about failing depaware check. Please, tell me what I am expected to do here.

Thank you.

Alex

bradfitz commented:
> I also don't know much about failing depaware check. Please, tell me what I am expected to do here.

See the top-level Makefile.

Run `make updatedeps`

alexbrainman commented:
> Run `make updatedeps`

I did. depaware step still fails:

```
The list of dependencies in /home/runner/work/tailscale/tailscale/cmd/tailscaled/depaware.txt is out of date.

--- before
+++ after
@@ -70,7 +70,7 @@
         tailscale.com/portlist                                       from tailscale.com/ipn
         tailscale.com/safesocket                                     from tailscale.com/ipn/ipnserver
         tailscale.com/smallzstd                                      from tailscale.com/ipn/ipnserver+
-        tailscale.com/syncs                                          from tailscale.com/net/interfaces+
+     💣 tailscale.com/syncs                                          from tailscale.com/net/interfaces+
         tailscale.com/tailcfg                                        from tailscale.com/control/controlclient+
   DW    tailscale.com/tempfork/osexec                                from tailscale.com/portlist
    W    tailscale.com/tsconst                                        from tailscale.com/net/interfaces
exit status 1
Error: Process completed with exit code 1.
```

I have no idea what the problem is. I will wait for you to tell me.

Thank you.

Alex

josharian commented on cmd/tailscale/depaware.txt:
```diff
@@ -64,7 +64,7 @@ tailscale.com/cmd/tailscale dependencies: (generated by github.com/tailscale/dep
         tailscale.com/paths                                          from tailscale.com/cmd/tailscale/cli
         tailscale.com/portlist                                       from tailscale.com/ipn
         tailscale.com/safesocket                                     from tailscale.com/cmd/tailscale/cli
-     💣 tailscale.com/syncs                                          from tailscale.com/net/interfaces+
```
Try reverting this line.

josharian commented on cmd/tailscaled/depaware.txt:
```diff
@@ -70,7 +70,7 @@ tailscale.com/cmd/tailscaled dependencies: (generated by github.com/tailscale/de
         tailscale.com/portlist                                       from tailscale.com/ipn
         tailscale.com/safesocket                                     from tailscale.com/ipn/ipnserver
         tailscale.com/smallzstd                                      from tailscale.com/ipn/ipnserver+
-     💣 tailscale.com/syncs                                          from tailscale.com/net/interfaces+
```
And this line.

alexbrainman replied:
Done.

alexbrainman reviewed (commented):
Hopefully I did not make any mistakes.

Alex

josharian commented:
depaware is happy now. :) Leaving for Brad.

alexbrainman commented:
> depaware is happy now. :)

Why did not command

`make updatedeps`

work without my manual adjustments? Do you want me to debug something here?

Alex

bradfitz commented:
`make updatedeps` doesn't update git state.

alexbrainman commented:
> `make updatedeps` doesn't update git state.

What I did is

```
make updatedeps
git add .
git commit --amend --no-edit
```

That obviously did not work, because I had to adjust my commit manually.

What did I do wrong?

Thank you.

Alex

bradfitz commented:
That should've worked. Maybe it did but you pushed something else to GitHub? No clue.

alexbrainman commented:
> That should've worked.

It is still broken here:

```
a@spot: oss $ git rev-parse --short HEAD
02526af
a@spot: oss $ git status
On branch alexbrainman/update_winipcfg
Your branch is up to date with 'origin/alexbrainman/update_winipcfg'.

nothing to commit, working tree clean
a@spot: oss $ make updatedeps 
go run github.com/tailscale/depaware --update tailscale.com/cmd/tailscaled
go run github.com/tailscale/depaware --update tailscale.com/cmd/tailscale
a@spot: oss $ git diff
diff --git a/cmd/tailscale/depaware.txt b/cmd/tailscale/depaware.txt
index 7403a1e..5252f31 100644
--- a/cmd/tailscale/depaware.txt
+++ b/cmd/tailscale/depaware.txt
@@ -64,7 +64,7 @@ tailscale.com/cmd/tailscale dependencies: (generated by github.com/tailscale/dep
         tailscale.com/paths                                          from tailscale.com/cmd/tailscale/cli
         tailscale.com/portlist                                       from tailscale.com/ipn
         tailscale.com/safesocket                                     from tailscale.com/cmd/tailscale/cli
-     <U+1F4A3> tailscale.com/syncs                                          from tailscale.com/net/interfaces+
+        tailscale.com/syncs                                          from tailscale.com/net/interfaces+
         tailscale.com/tailcfg                                        from tailscale.com/cmd/tailscale/cli+
   DW    tailscale.com/tempfork/osexec                                from tailscale.com/portlist
    W    tailscale.com/tsconst                                        from tailscale.com/net/interfaces
diff --git a/cmd/tailscaled/depaware.txt b/cmd/tailscaled/depaware.txt
index 319cf47..63d6821 100644
--- a/cmd/tailscaled/depaware.txt
+++ b/cmd/tailscaled/depaware.txt
@@ -70,7 +70,7 @@ tailscale.com/cmd/tailscaled dependencies: (generated by github.com/tailscale/de
         tailscale.com/portlist                                       from tailscale.com/ipn
         tailscale.com/safesocket                                     from tailscale.com/ipn/ipnserver
         tailscale.com/smallzstd                                      from tailscale.com/ipn/ipnserver+
-     <U+1F4A3> tailscale.com/syncs                                          from tailscale.com/net/interfaces+
+        tailscale.com/syncs                                          from tailscale.com/net/interfaces+
         tailscale.com/tailcfg                                        from tailscale.com/control/controlclient+
   DW    tailscale.com/tempfork/osexec                                from tailscale.com/portlist
    W    tailscale.com/tsconst                                        from tailscale.com/net/interfaces
a@spot: oss $
```

Do you want me to debug this?

I use `go version devel +e59ca9d Sat Aug 22 10:23:03 2020 +1000 linux/amd64`. I also use oss submodule of winmin repo, if it matters.

Alex

bradfitz commented:
> Do you want me to debug this?

No. Sounds like git submodule confusion.

alexbrainman commented:
> I sent [WireGuard/wireguard-windows#9](https://github.com/WireGuard/wireguard-windows/pull/9) upstream. Let's get that in first so we're not losing anything with this change.

@bradfitz this was merged

https://github.com/WireGuard/wireguard-windows/commit/dd60d0447f8149ca74297471fb44b92ae83a1d6b

Unfortunately the latest version of golang.zx2c4.com/wireguard/windows is v0.1.1, and your commit is not part of that version. Should we keep waiting?

Alex

bradfitz commented:
> Unfortunately the latest version of golang.zx2c4.com/wireguard/windows is v0.1.1, and your commit is not part of that version. Should we keep waiting?

You don't need to use a tagged version. Feel free to make our go.mod use a specific git commit (`go get ...@master`)

alexbrainman commented:
> Feel free to make our go.mod use a specific git commit (`go get ...@master`)

Done. PTAL.

Alex

bradfitz commented on wgengine/router/ifconfig_windows.go:
```diff
@@ -588,28 +640,46 @@ func deltaRouteData(a, b []*winipcfg.RouteData) (add, del []*winipcfg.RouteData)
+	i := 0
+	for _, route := range routes {
+		if route.InterfaceLUID == ifc.LUID {
+			matches[i] = &route
```
routes is of type `[]T` (https://godoc.org/golang.zx2c4.com/wireguard/windows/tunnel/winipcfg#GetIPForwardTable2) not `[]*T`, and this is a range over that value type, so `&route` is the same value as you iterate as you range over the slice.

alexbrainman replied:
Oops. Fixed.

I was doing this change pretty much mechanically, without much understanding of what I was looking at. Go compiler was guiding me along. You can blame the compiler.

PTAL

Alex

bradfitz commented:
I cherry-picked this in (and fixed the trivial go.mod merge locally), but then it doesn't start up:

```
srv: 3.1M/0.0M wgengine.NewUserspaceEngine: (try 2) monitorDefaultRoutes, after 13ms: Element not found.
...
srv: 4.7M/0.0M wgengine.NewUserspaceEngine: (try 4) monitorDefaultRoutes, after 13ms: Element not found.
...
srv: 4.6M/0.0M Bringing router up...
srv: 4.6M/0.0M router: after 180ms, removed firewall rule (wasPresent=false)
srv: 4.7M/0.0M Device closing
srv: 4.7M/0.0M Routine: event worker - stopped
srv: 4.7M/0.0M Interface closed
srv: 4.7M/0.0M 1: getEngine failed again: wgengine.NewUserspaceEngine: monitorDefaultRoutes, after 14ms: Element not found.
```

etc

alexbrainman commented:
> I cherry-picked this in (and fixed the trivial go.mod merge locally), but then it doesn't start up:

Oops. I did not even test this here. Sorry about that.

I will debug this over the weekend. Hopefully it is something simple.

Alex

alexbrainman commented:
> srv: 3.1M/0.0M wgengine.NewUserspaceEngine: (try 2) monitorDefaultRoutes, after 13ms: Element not found.

I cannot reproduce this error here. It runs fine here. And the `Element not found` error message is not good. Please try to debug monitorDefaultRoutes to see where it fails. I updated my change on top of current main. I will push in a minute.

Thank you.

Alex

alexbrainman commented:
> I cannot reproduce this error here. It runs fine here. And the `Element not found` error message is not good. Please try to debug monitorDefaultRoutes to see where it fails. I updated my change on top of current main.

Pinging @bradfitz - please, debug this so we can fix this change and submit. Before it goes stale again.

Thank you.

Alex

bradfitz commented:
It was the missing v6 error type. Fixed:

https://github.com/tailscale/tailscale/compare/f0891faa27a19b7e396230594bc8196b64ec01ab..8fea8dff5786d116cdb29d7806c53d3cb3f6e276#diff-84f6895565ed0043f2a7c9cb289741239aca9eed5c0eab819effe805c2d34499L77

alexbrainman commented:
> It was the missing v6 error type. Fixed:

Cool.

The only problem is, you removed isMissingIPv6Err function, and replaced it with `errors.Is(err, windows.ERROR_NOT_FOUND)`. isMissingIPv6Err had a nice comment explaining what the function does, and now you deleted that comment.

Alex
